File: app/src/main/java/jp/juggler/subwaytooter/ActColumnList.java
package jp.juggler.subwaytooter;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.woxthebox.draglistview.DragItem;
import com.woxthebox.draglistview.DragItemAdapter;
import com.woxthebox.draglistview.DragListView;
import com.woxthebox.draglistview.swipe.ListSwipeHelper;
import com.woxthebox.draglistview.swipe.ListSwipeItem;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import jp.juggler.subwaytooter.util.Utils;

public class ActColumnList extends AppCompatActivity {
	public static final String EXTRA_ORDER = "order";
	public static final String EXTRA_SELECTION = "selection";
	
	@Override
	protected void onCreate( @Nullable Bundle savedInstanceState ){
		super.onCreate( savedInstanceState );
		App1.setActivityTheme( this, false );
		initUI();
		
		if( savedInstanceState != null ){
			restoreData(
				savedInstanceState.getString( EXTRA_ORDER )
				, savedInstanceState.getInt( EXTRA_SELECTION )
			);
		}else{
			Intent intent = getIntent();
			restoreData(
				intent.getStringExtra( EXTRA_ORDER )
				, intent.getIntExtra( EXTRA_SELECTION, - 1 )
			);
		}
	}
	
	@Override
	protected void onSaveInstanceState( Bundle outState ){
		super.onSaveInstanceState( outState );
		//
		outState.putInt( EXTRA_SELECTION, old_selection );
		//
		JSONArray array = new JSONArray();
		List< MyItem > item_list = listAdapter.getItemList();
		for( int i = 0, ie = item_list.size() ; i < ie ; ++ i ){
			array.put( item_list.get( i ).json );
		}
		outState.putString( EXTRA_ORDER, array.toString() );
	}
	
	@Override
	public void onBackPressed(){
		makeResult( - 1 );
		super.onBackPressed();
	}
	
	DragListView listView;
	MyListAdapter listAdapter;
	int old_selection;
	
	private void initUI(){
		setContentView( R.layout.act_column_list );
		
		// リストのアダプター
		listAdapter = new MyListAdapter();
		
		// ハンドル部分をドラッグで並べ替えできるRecyclerView
		listView = (DragListView) findViewById( R.id.drag_list_view );
		listView.setLayoutManager( new LinearLayoutManager( this ) );
		listView.setAdapter( listAdapter, true );
		listView.setCanDragHorizontally( false );
		listView.setCustomDragItem( new MyDragItem( this, R.layout.lv_column_list ) );
		
		listView.getRecyclerView().setVerticalScrollBarEnabled( true );
		listView.setDragListListener( new DragListView.DragListListenerAdapter() {
			@Override
			public void onItemDragStarted( int position ){
				// 操作中はリフレッシュ禁止
				// mRefreshLayout.setEnabled( false );
			}
			
			@Override
			public void onItemDragEnded( int fromPosition, int toPosition ){
				// 操作完了でリフレッシュ許可
				// mRefreshLayout.setEnabled( USE_SWIPE_REFRESH );

//				if( fromPosition != toPosition ){
//					// 並べ替えが発生した
//				}
			}
		} );
		
		// リストを左右スワイプした
		listView.setSwipeListener( new ListSwipeHelper.OnSwipeListenerAdapter() {
			
			@Override
			public void onItemSwipeStarted( ListSwipeItem item ){
				// 操作中はリフレッシュ禁止
				// mRefreshLayout.setEnabled( false );
			}
			
			@Override
			public void onItemSwipeEnded( ListSwipeItem item, ListSwipeItem.SwipeDirection swipedDirection ){
				// 操作完了でリフレッシュ許可
				// mRefreshLayout.setEnabled( USE_SWIPE_REFRESH );
				
				// 左にスワイプした(右端に青が見えた) なら要素を削除する
				if( swipedDirection == ListSwipeItem.SwipeDirection.LEFT ){
					MyItem adapterItem = (MyItem) item.getTag();
					if( adapterItem.json.optBoolean( Column.KEY_DONT_CLOSE, false ) ){
						Utils.showToast( ActColumnList.this, false, R.string.column_has_dont_close_option );
						listView.resetSwipedViews( null );
						return;
					}
					listAdapter.removeItem( listAdapter.getPositionForItem( adapterItem ) );
				}
			}
		} );
	}
	
	void restoreData( String svColumnList, int ivSelection ){
		
		this.old_selection = ivSelection;
		
		ArrayList< MyItem > tmp_list = new ArrayList<>();
		try{
			JSONArray array = new JSONArray( svColumnList );
			for( int i = 0, ie = array.length() ; i < ie ; ++ i ){
				try{
					JSONObject src = array.optJSONObject( i );
					MyItem item = new MyItem( src, i, this );
					if( src != null ){
						tmp_list.add( item );
						if( old_selection == item.old_index ){
							item.setOldSelection( true );
						}
					}
				}catch( Throwable ex2 ){
					ex2.printStackTrace();
				}
			}
		}catch( Throwable ex ){
			ex.printStackTrace();
		}
		listAdapter.setItemList( tmp_list );
	}
	
	void makeResult( int new_selection ){
		Intent intent = new Intent();
		
		List< MyItem > item_list = listAdapter.getItemList();
		// どの要素を選択するか
		if( new_selection >= 0 && new_selection < listAdapter.getItemCount() ){
			intent.putExtra( EXTRA_SELECTION, new_selection );
		}else{
			for( int i = 0, ie = item_list.size() ; i < ie ; ++ i ){
				if( item_list.get( i ).bOldSelection ){
					intent.putExtra( EXTRA_SELECTION, i );
					break;
				}
			}
		}
		// 並べ替え用データ
		ArrayList< Integer > order_list = new ArrayList<>();
		for( MyItem item : item_list ){
			order_list.add( item.old_index );
		}
		intent.putExtra( EXTRA_ORDER, order_list );
		
		setResult( RESULT_OK, intent );
	}
	
	private void performItemSelected( MyItem item ){
		int idx = listAdapter.getPositionForItem( item );
		makeResult( idx );
		finish();
	}
	
	// リスト要素のデータ
	static class MyItem {
		long id;
		JSONObject json;
		String name;
		String acct;
		int acct_color_fg;
		int acct_color_bg;
		boolean bOldSelection;
		int old_index;
		int type;
		
		MyItem( JSONObject src, long id, Context context ){
			this.json = src;
			this.name = src.optString( Column.KEY_COLUMN_NAME );
			this.acct = src.optString( Column.KEY_COLUMN_ACCESS );
			int c = src.optInt( Column.KEY_COLUMN_ACCESS_COLOR, 0 );
			this.acct_color_fg = c != 0 ? c : Styler.getAttributeColor( context, R.attr.colorColumnListItemText );
			c = src.optInt( Column.KEY_COLUMN_ACCESS_COLOR_BG, 0 );
			this.acct_color_bg = c;
			this.old_index = src.optInt( Column.KEY_OLD_INDEX );
			this.id = id;
			this.type = src.optInt( Column.KEY_TYPE );
		}
		
		void setOldSelection( boolean b ){
			bOldSelection = b;
		}
	}
	
	// リスト要素のViewHolder
	class MyViewHolder extends DragItemAdapter.ViewHolder {
		final View ivBookmark;
		final TextView tvAccess;
		final TextView tvName;
		final ImageView ivColumnIcon;
		final int acct_pad_lr;
		
		MyViewHolder( final View viewRoot ){
			super( viewRoot
				, R.id.ivDragHandle // View ID。 ここを押すとドラッグ操作をすぐに開始する
				, true // 長押しでドラッグ開始するなら真
			);
			
			ivBookmark = viewRoot.findViewById( R.id.ivBookmark );
			tvAccess = (TextView) viewRoot.findViewById( R.id.tvAccess );
			tvName = (TextView) viewRoot.findViewById( R.id.tvName );
			ivColumnIcon = (ImageView) viewRoot.findViewById( R.id.ivColumnIcon );
			acct_pad_lr = (int) ( 0.5f + 4f * viewRoot.getResources().getDisplayMetrics().density );
			
			// リスト要素のビューが ListSwipeItem だった場合、Swipe操作を制御できる
			if( viewRoot instanceof ListSwipeItem ){
				ListSwipeItem lsi = (ListSwipeItem) viewRoot;
				lsi.setSwipeInStyle( ListSwipeItem.SwipeInStyle.SLIDE );
				lsi.setSupportedSwipeDirection( ListSwipeItem.SwipeDirection.LEFT );
			}
			
		}
		
		void bind( MyItem item ){
			itemView.setTag( item ); // itemView は親クラスのメンバ変数
			ivBookmark.setVisibility( item.bOldSelection ? View.VISIBLE : View.INVISIBLE );
			tvAccess.setText( item.acct );
			tvAccess.setTextColor( item.acct_color_fg );
			tvAccess.setBackgroundColor( item.acct_color_bg );
			tvAccess.setPaddingRelative( acct_pad_lr, 0, acct_pad_lr, 0 );
			tvName.setText( item.name );
			ivColumnIcon.setImageResource( Styler.getAttributeResourceId(
				ActColumnList.this, Column.getIconAttrId( item.type ) ) );
		}

//		@Override
//		public boolean onItemLongClicked( View view ){
//			return false;
//		}
		
		@Override
		public void onItemClicked( View view ){
			MyItem item = (MyItem) itemView.getTag(); // itemView は親クラスのメンバ変数
			ActColumnList activity = ( (ActColumnList) Utils.getActivityFromView( view ) );
			if( activity != null ) activity.performItemSelected( item );
		}
	}
	
	// ドラッグ操作中のデータ
	private class MyDragItem extends DragItem {
		MyDragItem( Context context, int layoutId ){
			super( context, layoutId );
		}
		
		@Override
		public void onBindDragView( View clickedView, View dragView ){
			MyItem item = (MyItem) clickedView.getTag();
			
			TextView tv = (TextView) dragView.findViewById( R.id.tvAccess );
			tv.setText( item.acct );
			tv.setTextColor( item.acct_color_fg );
			tv.setBackgroundColor( item.acct_color_bg );
			
			tv = (TextView) dragView.findViewById( R.id.tvName );
			tv.setText( item.name );
			
			ImageView ivColumnIcon = (ImageView) dragView.findViewById( R.id.ivColumnIcon );
			ivColumnIcon.setImageResource( Styler.getAttributeResourceId(
				ActColumnList.this, Column.getIconAttrId( item.type ) ) );
			
			dragView.findViewById( R.id.ivBookmark ).setVisibility(
				clickedView.findViewById( R.id.ivBookmark ).getVisibility()
			);
			
			dragView.findViewById( R.id.item_layout ).setBackgroundColor(
				Styler.getAttributeColor( ActColumnList.this, R.attr.list_item_bg_pressed_dragged )
			);
		}
	}
	
	private class MyListAdapter extends DragItemAdapter< MyItem, MyViewHolder > {
		
		MyListAdapter(){
			super();
			setHasStableIds( true );
			setItemList( new ArrayList< MyItem >() );
		}
		
		@Override
		public MyViewHolder onCreateViewHolder( ViewGroup parent, int viewType ){
			View view = getLayoutInflater().inflate( R.layout.lv_column_list, parent, false );
			return new MyViewHolder( view );
		}
		
		@Override
		public void onBindViewHolder( MyViewHolder holder, int position ){
			super.onBindViewHolder( holder, position );
			holder.bind( getItemList().get( position ) );
		}
		
		@Override
		public long getItemId( int position ){
			MyItem item = mItemList.get( position ); // mItemList は親クラスのメンバ変数
			return item.id;
		}
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/ActColumnList.kt
package jp.juggler.subwaytooter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import com.woxthebox.draglistview.DragItem
import com.woxthebox.draglistview.DragItemAdapter
import com.woxthebox.draglistview.DragListView
import com.woxthebox.draglistview.swipe.ListSwipeHelper
import com.woxthebox.draglistview.swipe.ListSwipeItem
import jp.juggler.subwaytooter.api.entity.Acct
import jp.juggler.subwaytooter.api.showApiError
import jp.juggler.subwaytooter.column.ColumnEncoder
import jp.juggler.subwaytooter.column.ColumnType
import jp.juggler.subwaytooter.databinding.ActColumnListBinding
import jp.juggler.subwaytooter.databinding.LvColumnListBinding
import jp.juggler.subwaytooter.dialog.DlgConfirm.confirm
import jp.juggler.util.backPressed
import jp.juggler.util.coroutine.launchMain
import jp.juggler.util.data.JsonObject
import jp.juggler.util.data.notZero
import jp.juggler.util.data.toJsonArray
import jp.juggler.util.int
import jp.juggler.util.log.LogCategory
import jp.juggler.util.ui.attrColor
import jp.juggler.util.ui.setContentViewAndInsets
import jp.juggler.util.ui.setNavigationBack
import jp.juggler.util.ui.vg
import org.jetbrains.anko.backgroundColor
import org.jetbrains.anko.textColor

class ActColumnList : AppCompatActivity() {

    companion object {

        private val log = LogCategory("ActColumnList")
        private const val TMP_FILE_COLUMN_LIST = "tmp_column_list"

        // リザルトに使うのでpublic
        const val EXTRA_ORDER = "order"
        const val EXTRA_SELECTION = "selection"

        fun createIntent(activity: ActMain, currentItem: Int) =
            Intent(activity, ActColumnList::class.java).apply {
                val array = activity.appState.encodeColumnList()
                AppState.saveColumnList(activity, TMP_FILE_COLUMN_LIST, array)
                putExtra(EXTRA_SELECTION, currentItem)
            }
    }

    private val views by lazy {
        ActColumnListBinding.inflate(layoutInflater)
    }

    private val listAdapter by lazy { MyListAdapter() }

    private val defaultAcctColorFg by lazy {
        attrColor(R.attr.colorColumnHeaderAcct)
    }
    private val defaultColumnColorFg by lazy {
        attrColor(R.attr.colorColumnHeaderName)
    }
    private var oldSelection: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        backPressed {
            makeResult(-1)
            finish()
        }

        super.onCreate(savedInstanceState)

        App1.setActivityTheme(this)
        setContentViewAndInsets(views.root)
        initUI()

        if (savedInstanceState != null) {
            restoreData(savedInstanceState.int(EXTRA_SELECTION) ?: -1)
        } else {
            val intent = intent
            restoreData(intent?.int(EXTRA_SELECTION) ?: -1)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(EXTRA_SELECTION, oldSelection)
        val array = listAdapter.itemList.map { it.json }.toJsonArray()
        AppState.saveColumnList(this, TMP_FILE_COLUMN_LIST, array)
    }

    private fun initUI() {
        setSupportActionBar(views.toolbar)
        setNavigationBack(views.toolbar)
        fixHorizontalMargin(views.listView)

        // ハンドル部分をドラッグで並べ替えできるRecyclerView
        views.listView.setLayoutManager(androidx.recyclerview.widget.LinearLayoutManager(this))
        views.listView.setAdapter(listAdapter, true)
        views.listView.setCanDragHorizontally(false)
        views.listView.setCustomDragItem(MyDragItem(this))

        views.listView.recyclerView.isVerticalScrollBarEnabled = true
        views.listView.setDragListListener(object : DragListView.DragListListenerAdapter() {
            override fun onItemDragStarted(position: Int) = Unit
            override fun onItemDragEnded(fromPosition: Int, toPosition: Int) = Unit
        })

        // リストを左右スワイプした
        views.listView.setSwipeListener(object : ListSwipeHelper.OnSwipeListenerAdapter() {
            override fun onItemSwipeStarted(item: ListSwipeItem) = Unit
            override fun onItemSwipeEnded(
                item: ListSwipeItem,
                swipedDirection: ListSwipeItem.SwipeDirection?,
            ) {
                // 左にスワイプした(右端に青が見えた) なら要素を削除する
                if (swipedDirection == ListSwipeItem.SwipeDirection.LEFT) {
                    val adapterItem = (item.tag as MyViewHolder).lastItem ?: return
                    launchMain {
                        try {
                            if (adapterItem.json.optBoolean(ColumnEncoder.KEY_DONT_CLOSE, false)) {
                                confirm(R.string.confirm_remove_column_mark_as_dont_close)
                            }
                            listAdapter.removeItem(listAdapter.getPositionForItem(adapterItem))
                        } catch (ex: Throwable) {
                            showApiError(ex)
                        } finally {
                            try {
                                views.listView.resetSwipedViews(null)
                            } catch (_: Throwable) {
                            }
                        }
                    }
                }
            }
        })
    }

    private fun restoreData(ivSelection: Int) {
        oldSelection = ivSelection
        // DragItemAdapter はMutableListを要求する
        listAdapter.itemList = ArrayList<MyItem>().apply {
            try {
                AppState.loadColumnList(applicationContext, TMP_FILE_COLUMN_LIST)
                    ?.objectList()
                    ?.forEachIndexed { index, src ->
                        try {
                            val item = MyItem(
                                src,
                                index.toLong(),
                            )
                            add(item)
                            if (oldSelection == item.oldIndex) {
                                item.setOldSelection(true)
                            }
                        } catch (ex: Throwable) {
                            log.e(ex, "restoreData: item decode failed.")
                        }
                    }
            } catch (ex: Throwable) {
                log.e(ex, "restoreData failed.")
            }
        }
    }

    private fun makeResult(newSelection: Int) {
        val intent = Intent()

        val itemList = listAdapter.itemList
        // どの要素を選択するか
        if (newSelection >= 0 && newSelection < listAdapter.itemCount) {
            intent.putExtra(EXTRA_SELECTION, newSelection)
        } else {
            var i = 0
            val ie = itemList.size
            while (i < ie) {
                if (itemList[i].bOldSelection) {
                    intent.putExtra(EXTRA_SELECTION, i)
                    break
                }
                ++i
            }
        }
        // 並べ替え用データ
        val orderList = ArrayList<Int>()
        for (item in itemList) {
            orderList.add(item.oldIndex)
        }
        intent.putExtra(EXTRA_ORDER, orderList)

        setResult(Activity.RESULT_OK, intent)
    }

    private fun performItemSelected(item: MyItem?) {
        item ?: return
        val idx = listAdapter.getPositionForItem(item)
        makeResult(idx)
        finish()
    }

    // リスト要素のデータ
    internal inner class MyItem(
        val json: JsonObject,
        val id: Long,
        val name: String = json.optString(ColumnEncoder.KEY_COLUMN_NAME),
        val acct: Acct = Acct.parse(json.optString(ColumnEncoder.KEY_COLUMN_ACCESS_ACCT)),
        val acctName: String = json.optString(ColumnEncoder.KEY_COLUMN_ACCESS_STR),
        val oldIndex: Int = json.optInt(ColumnEncoder.KEY_OLD_INDEX),
        val type: ColumnType = ColumnType.parse(json.optInt(ColumnEncoder.KEY_TYPE)),
        val acctColorBg: Int = json.optInt(ColumnEncoder.KEY_COLUMN_ACCESS_COLOR_BG, 0),
        val acctColorFg: Int = json.optInt(ColumnEncoder.KEY_COLUMN_ACCESS_COLOR, 0),
        val columnColorFg: Int = json.optInt(ColumnEncoder.KEY_HEADER_TEXT_COLOR, 0),
        val columnColorBg: Int = json.optInt(ColumnEncoder.KEY_HEADER_BACKGROUND_COLOR, 0),
    ) {
        var bOldSelection: Boolean = false

        fun setOldSelection(b: Boolean) {
            bOldSelection = b
        }
    }

    // リスト要素のViewHolder
    private inner class MyViewHolder(
        parent: ViewGroup?,
        val views: LvColumnListBinding =
            LvColumnListBinding.inflate(layoutInflater, parent, false),
    ) : DragItemAdapter.ViewHolder(
        views.root,
        R.id.ivDragHandle, // View ID。 ここを押すとドラッグ操作をすぐに開始する
        true, // 長押しでドラッグ開始するなら真
    ) {
        var lastItem: MyItem? = null
        val acctPadLr = (0.5f + 4f * views.root.resources.displayMetrics.density).toInt()

        init {
            views.root.tag = this
            views.root.setSwipeInStyle(ListSwipeItem.SwipeInStyle.SLIDE)
            views.root.supportedSwipeDirection = ListSwipeItem.SwipeDirection.LEFT
        }

        fun bind(item: MyItem?) {
            item ?: return
            lastItem = item
            views.ivSelected.vg(item.bOldSelection)
            views.tvAccess.text = item.acctName
            views.tvAccess.setTextColor(item.acctColorFg.notZero() ?: defaultAcctColorFg)
            views.tvAccess.setBackgroundColor(item.acctColorBg)
            views.tvAccess.setPaddingRelative(acctPadLr, 0, acctPadLr, 0)

            val columnColorFg = item.columnColorFg.notZero() ?: defaultColumnColorFg
            views.llColumn.backgroundColor = item.columnColorBg
            views.tvColumnName.text = item.name
            views.tvColumnName.textColor = columnColorFg
            views.ivColumnIcon.setImageResource(item.type.iconId(item.acct))
            views.ivColumnIcon.imageTintList = ColorStateList.valueOf(columnColorFg)
            views.ivSelected.imageTintList = ColorStateList.valueOf(columnColorFg)

            // 背景色がテーマ次第なので、カラム設定の色を反映するとアイコンが見えなくなる可能性がある
            // よってアイコンやテキストにカラム設定の色を反映しない
        }

        override fun onItemClicked(view: View?) {
            performItemSelected(lastItem)
        }
    }

    // ドラッグ操作中のデータ
    private inner class MyDragItem(context: Context) : DragItem(context, R.layout.lv_column_list) {
        override fun onBindDragView(clickedView: View, dragView: View) {
            val dragViews = LvColumnListBinding.bind(dragView)
            val clickVh = clickedView.tag as MyViewHolder
            val clickViews = clickVh.views
            val item = clickVh.lastItem!!

            dragViews.tvAccess.run {
                text = item.acctName
                textColor = item.acctColorFg.notZero() ?: defaultAcctColorFg
                backgroundColor = item.acctColorBg
            }
            val columnColorFg = item.columnColorFg.notZero() ?: defaultColumnColorFg
            dragViews.tvColumnName.run {
                text = item.name
                textColor = columnColorFg
            }
            dragViews.ivColumnIcon.imageTintList = ColorStateList.valueOf(columnColorFg)
            dragViews.ivSelected.imageTintList = ColorStateList.valueOf(columnColorFg)
            dragViews.llColumn.backgroundColor = item.columnColorBg
            dragViews.ivColumnIcon.setImageResource(item.type.iconId(item.acct))
            dragViews.ivSelected.visibility = clickViews.ivSelected.visibility
            dragViews.itemLayout.setBackgroundColor(attrColor(R.attr.list_item_bg_pressed_dragged))
        }
    }

    private inner class MyListAdapter : DragItemAdapter<MyItem, MyViewHolder>() {

        init {
            setHasStableIds(true)
            itemList = ArrayList()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder =
            MyViewHolder(parent)

        override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
            super.onBindViewHolder(holder, position)
            holder.bind(itemList[position])
        }

        override fun getUniqueItemId(position: Int): Long {
            val item = mItemList[position] // mItemList は親クラスのメンバ変数
            return item.id
        }
    }
}
